src/clj/puppetlabs/http/client/async.clj

```
@@ -326,39 +137,80 @@
-                                        (.failed future-callback ex))
-                                      (cancelled [this]
-                                        (.setFinalResult consumer nil)
-                                        (.cancelled future-callback)))]
-    (.execute client
-              (HttpAsyncMethods/create request)
-              consumer
-              streaming-complete-callback)))
+   response :- common/Response]
+  (if callback
+    (try
+      (callback response)
+      (catch Exception e
+        {:opts opts :error e}))
+    response))
+
+(schema/defn java-content-type->clj :- common/ContentType
+  [java-content-type]
+  (if java-content-type
+    {:mime-type (.getMimeType java-content-type)
+     :charset   (.getCharset java-content-type)}))
+
+(schema/defn get-response-delivery-delegate :- ResponseDeliveryDelegate
+  [opts :- common/UserRequestOptions
+   result :- common/ResponsePromise]
+  (reify ResponseDeliveryDelegate
+    (deliverResponse
+      [_ _ orig-encoding body headers status content-type callback]
+      (let [response {:opts                  opts
+                      :orig-content-encoding orig-encoding
+                      :status                status
+                      :headers               (into {} headers)
+                      :content-type          (java-content-type->clj content-type)
+                      :body                  body}
+            response (callback-response opts callback response)]
+        (deliver result response)))
+    (deliverResponse
+      [_ _ e]
+      (deliver result {:opts opts :error e}))))
+
+(schema/defn clojure-method->java
+  [opts :- common/UserRequestOptions]
+  (case (:method opts)
+    :delete HttpMethod/DELETE
+    :get HttpMethod/GET
+    :head HttpMethod/HEAD
+    :options HttpMethod/OPTIONS
+    :patch HttpMethod/PATCH
+    :post HttpMethod/POST
+    :put HttpMethod/PUT
+    :trace HttpMethod/TRACE))
```

I noticed that previously we would throw an error if the method wasn't found (https://github.com/puppetlabs/clj-http-client/pull/47/files#diff-f2e3a844d8f57439d128ef47be66d78dL163), but with this now we don't. Should we throw an error with a better message (otherwise it would be `java.lang.IllegalArgumentException: No matching clause:`)? Is that even a case that we can reach? Does it matter that much? (there's no test I can find)